buildspec.yml

```diff
@@ -0,0 +1,33 @@
+version: 0.2
+
+phases:
+  build:
+    commands:
+      - git clone https://github.com/crungehottman/exodus-lambda.git
+      - cd exodus-lambda
+      - git checkout codebuild
+      - gpg --import .whitelist.gpg
```

Comment:
We could use the whitelist from a previous tag. Though that doesn't offer much more protection. Someone could just push two tagged commits, the first containing their key.

Originally, I thought of importing keys from distribution-password-store, but that would require cloning yet another repo and traversing the pubkey files. Anyway, distribution-password-store could potentially be hijacked too, right?

How many layers are enough?

Would importing from a key server be better somehow? We'd still need to include trusted email addresses somewhere for that, so it would seem to carry the same flaws as the current approach.

Reply:
> We could use the whitelist from a previous tag. Though that doesn't offer much more protection. Someone could just push two tagged commits, the first containing their key.

+1

Another potential approach: Instead of pulling the whitelist directly from git, the whitelist could exist as an artifact in our exodus-pipeline-artifacts bucket. CodeBuild would use the whitelist as a source artifact provided to the build to be used to import and verify the GPG keys. The whitelist from the git tag would exit the build phase as a build artifact. If the key used to sign the tag is in the whitelist, the build passes and the pipeline continues into a new 'replace whitelist' phase, which replaces the whitelist in the bucket with the whitelist from the git tag. If the GPG key used to sign the git tag is not in the whitelist, the build fails and the rest of the pipeline will not execute, including the replacement of the gpg whitelist -- the list from the previous successful deployment is maintained.

Reply:
> Originally, I thought of importing keys from distribution-password-store, but that would require cloning yet another repo and traversing the pubkey files. Anyway, distribution-password-store could potentially be hijacked too, right?

I don't consider that repo being hijacked as in scope as it is Red Hat internal.  However I don't think you can use it for that exact reason.  Also, based on the other conversation thread about how the buildspec changes get deployed, if you add such checks in the buildspec, they can simply be removed by the attacker, right?  Same goes for *any* verification added in the buildspec for that matter.

The threat model I want us to be robust against has assumptions:

1. Assume that *this repo* can be written to by an attacker
2. Assume anything already deployed in AWS is not compromised (if it was then the game is already lost)
3. Assume Red Hat internal repos & access control mechanisms are not compromised (though I'm not sure it's relevant)

Now imagine we have a CodePipeline deployed in AWS.  I read Nathan's comments in another thread as meaning: when that pipeline is used to deploy revision X, the build runs whatever build steps are in the buildspec file as of revision X.  That makes it futile to use the buildspec to verify what's being deployed, because the buildspec itself can be written to by an attacker.

If this understanding of the buildspec is correct then it seems like the design from RHELDST-20 asking to have GPG verification in the buildspec is wrong, it can't be done usefully here.  It could possibly be done usefully earlier in the pipeline though, but that also depends on how the pipeline is deployed.  This would be a lot easier to judge if we already had the pipeline to review as well.

Does this make sense?  If so I would recommend stripping out the GPG verification stuff here and instead shifting that  into the scope of RHELDST-25.